axelor-purchase/src/main/java/com/axelor/apps/purchase/web/PurchaseOrderLineController.java

```diff
@@ -221,25 +245,77 @@ public void getDiscount(ActionRequest request, ActionResponse response) {
     }
 
     try {
-      BigDecimal price = purchaseOrderLine.getPrice();
-
-      Map<String, Object> discounts =
-          purchaseOrderLineService.getDiscount(purchaseOrder, purchaseOrderLine, price);
+      Map<String, Object> discounts;
+      if (purchaseOrderLine.getProduct().getInAti()) {
+        discounts =
+            purchaseOrderLineService.getDiscount(
+                purchaseOrder, purchaseOrderLine, purchaseOrderLine.getInTaxPrice());
+      } else {
+        discounts =
+            purchaseOrderLineService.getDiscount(
+                purchaseOrder, purchaseOrderLine, purchaseOrderLine.getPrice());
+      }
 
       if (discounts != null) {
-
         response.setValue("discountAmount", discounts.get("discountAmount"));
         response.setValue("discountTypeSelect", discounts.get("discountTypeSelect"));
-
         if (discounts.get("price") != null) {
-          response.setValue("price", discounts.get("price"));
+          if (purchaseOrderLine.getProduct().getInAti()) {
+            response.setValue("inTaxPrice", discounts.get("price"));
+            response.setValue(
+                "price",
+                purchaseOrderLineService.convertUnitPrice(
+                    true, purchaseOrderLine.getTaxLine(), (BigDecimal) discounts.get("price")));
+          } else {
+            response.setValue("price", discounts.get("price"));
+            response.setValue(
+                "inTaxPrice",
+                purchaseOrderLineService.convertUnitPrice(
+                    false, purchaseOrderLine.getTaxLine(), (BigDecimal) discounts.get("price")));
+          }
         }
       }
     } catch (Exception e) {
       response.setFlash(e.getMessage());
     }
   }
 
+  /**
+   * Update the ex. tax unit price of an invoice line from its in. tax unit price.
+   *
+   * @param request
+   * @param response
+   */
+  public void updatePrice(ActionRequest request, ActionResponse response) {
+    Context context = request.getContext();
+
+    PurchaseOrderLine purchaseOrderLine = context.asType(PurchaseOrderLine.class);
+
+    BigDecimal inTaxPrice = purchaseOrderLine.getInTaxPrice();
+    TaxLine taxLine = purchaseOrderLine.getTaxLine();
+
+    response.setValue(
+        "price", purchaseOrderLineService.convertUnitPrice(true, taxLine, inTaxPrice));
```

Comment:
Everything in a controller should be in a try/catch block. When catching an exception, use `TracebackService` so the runtime exception can be saved.